stacks-signer/src/v0/signer_state.rs

```diff
@@ -38,6 +40,205 @@ use crate::signerdb::SignerDb;
 /// This is the latest supported protocol version for this signer binary
 pub static SUPPORTED_SIGNER_PROTOCOL_VERSION: u64 = 0;
 
+/// A struct used to determine the current global state
+#[derive(Debug)]
+pub struct GlobalStateEvaluator {
+    /// A mapping of signer addresses to their corresponding vote weight
+    address_weights: HashMap<StacksAddress, u32>,
+    /// A mapping of signer addresses to their corresponding updates
+    address_updates: HashMap<StacksAddress, StateMachineUpdateMessage>,
+    /// The total weight of all signers
+    total_weight: u32,
+}
+
+impl GlobalStateEvaluator {
+    /// Create a new state evaluator
+    pub fn new(
+        address_updates: HashMap<StacksAddress, StateMachineUpdateMessage>,
+        address_weights: HashMap<StacksAddress, u32>,
+    ) -> Self {
+        let total_weight = address_weights
+            .values()
+            .fold(0u32, |acc, val| acc.saturating_add(*val));
+        Self {
+            address_weights,
+            address_updates,
+            total_weight,
+        }
+    }
+
+    /// Determine what the active signer protocol version should be
+    ///
+    /// NOTE: do not call this function unless the evaluator has already been updated with the current local state
+    pub fn determine_active_signer_protocol_version(&self) -> Option<u64> {
+        let mut protocol_versions = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let entry = protocol_versions
+                .entry(update.local_supported_signer_protocol_version)
+                .or_insert_with(|| 0);
+            *entry += weight;
+            if *entry >= self.total_weight * 7 / 10 {
+                return Some(update.local_supported_signer_protocol_version);
+            }
+        }
```

Comment:
I think we want something more like:

```suggestion
        for (address, update) in &self.address_updates {
            let Some(weight) = self.address_weights.get(address) else {
                continue;
            };
            let entry = protocol_versions
                .entry(update.local_supported_signer_protocol_version)
                .or_insert_with(|| 0);
            *entry += weight;
        }
        // find the highest version number supported by a threshold number of signers
        let mut protocol_versions: Vec<_> = protocol_versions.into_iter().collect();
        protocol_versions.sort_by_key(|(version, _)| *version);
        let mut total_weight_support = 0;
        for (version, weight_support) in protocol_versions.into_iter().rev() {
            total_weight_support += weight_support;
            if total_weight_support >= self.total_weight * 7 / 10 {
                return Some(version)
            }
        }
        return None
```

Reply:
Right, I think the point is that if the system has like 3 equal weight signers, and their "highest support versions" are:

1, 1, 2.

We'd want the system to still be "okay", because the version 2 signer should be able to still communicate in version 1 (i.e., be backwards-compatible). So what we want is the highest version number that a threshold number of signers support. It's true that they may end up supporting multiple versions, but there's still guaranteed to be only one "highest supported" version.